Okay! I'll try to start from the beginning and explain everything.

The standard library has a feature called `panic_immediate_abort`. The idea is that if you're running on a system that can't emit panic messages or you don't care about them, you want the panic runtime optimized away. "The panic runtime" basically means this function and all the others like it: https://github.com/rust-lang/rust/blob/5b8bc568d28b2e922290c9a966b3231d0ce9398b/library/core/src/panicking.rs#L46-L52
and also all the formatting code they pull in.

Currently (I hope to change this eventually), we use `cfg_attr` so that by default the standard library puts `#[inline(never)]` and `#[cold]` on all the functions which are entrypoints to the panic runtime, because they are in fact cold, and also contain a lot of formatting code, so inlining them is undesirable. But if we don't want panic formatting in our build, we want calls to those functions optimized down to just a `ud2` (or equivalent). So when `-Zbuild-std-features=panic_immediate_abort` is passed, we swap out those attributes for `#[inline]` instead.

The LLVM `inlinehint` that `#[inline]` produces is not very important here, LLVM is trivially able to figure out that these functions are just an unreachable terminator. But since we want these functions to be inlined into our executable from the standard library, we need cross-crate-inlining. We _could_ require LTO, but it would be much nicer to not, and we can get the desired effect by swapping in just `#[inline]` on these functions.

---

With `#[inline(never)]` we eagerly codegen all the panic entrypoints when compiling the standard library. But when we swap to `#[inline]`, we only codegen of functions which are determined to be reachable, as determined by `MirUsedCollector` and CGU partitioning.

This can go wrong if `MirUsedCollector` doesn't realize that a function is used. I recently submitted https://github.com/rust-lang/rust/pull/118693 to fix a scenario much like this PR. In that other PR, since `panic_misaligned_pointer_dereference` is not actually called in MIR, we need to tell `MirUsedCollector` that a particular kind of MIR terminator causes that function to become reachable. This situation with `panic_misaligned_pointer_dereference` is what motivated me to create this PR. The root problem is that the compiler assumes that `MirUsedCollector` finds all the used MIR, so if it ever doesn't we have a lurking problem.

So. The problem pattern here is that we add `#[inline]` to a function then said function becomes an undefined symbol at link time, because said function requires special treatment in our reachability analysis that was missing. This PR is about finding and fixing those cases.

---

The best way to find these cases is to apply our `#[inline]` style codegen to everything. `-Zcross-crate-inline-threshold=always` exists to provide this, but when I initially implemented it I didn't realize that the codegen strategy and the inlining optimization are completely separate- we can have cross-crate codegen with inlining forbidden. So previously `-Zcross-crate-inline-threshold=always` meant to apply cross-crate codegen to all functions except those which are `#[no_mangle]` or `#[inline(never)]` and now it only `#[no_mangle]` functions are exempted.

So I made that change then tried to build the compiler with `-Zcross-crate-inline-threshold=always x build --stage 2` and got some undefined symbol errors that I started to debug.

The first errors were for `panic_nounwind`. This symbol has exactly the same problem as `panic_misaligned_pointer_dereference`; codegen inserts calls to it based on some other MIR construct, in this case the intrinsic validity assertions, which are declared here: https://github.com/rust-lang/rust/blob/028b6d152e904bbc02dc3cd67e4cbdffcbd039e1/library/core/src/intrinsics.rs#L1068-L1092 and the codegen for them is here: https://github.com/rust-lang/rust/blob/028b6d152e904bbc02dc3cd67e4cbdffcbd039e1/compiler/rustc_codegen_ssa/src/mir/block.rs#L682-L685. I basically combined the way codegen detects these intrinsics with how `MirUsedCollector` adds mono items.

With that fixed, I got an undefined reference to `lang_start`. That one was rather confusing because it's already special-cased in `MirUsedCollector` here: https://github.com/rust-lang/rust/blob/028b6d152e904bbc02dc3cd67e4cbdffcbd039e1/compiler/rustc_monomorphize/src/collector.rs#L1246-L1256 the problem is that though it gets collected, it gets dropped by CGU partitioning. We partition by `InstantiationMode::GloballyShared` items, then stick onto each CGU all the items that are reachable from the contents of the CGU. But if `lang_start` gets `InstantiationMode::LocalCopy` then it's not a root item and it's of course not called from anywhere. So CGU partitioning just forgets about it.

Phew that's a lot of text. Does this help?

--- a/compiler/rustc_mir_transform/src/cross_crate_inline.rs
+++ b/compiler/rustc_mir_transform/src/cross_crate_inline.rs
@@ -22,19 +22,24 @@ fn cross_crate_inlinable(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
         return false;
     }
 
+    // This just reproduces the logic from Instance::requires_inline.
+    match tcx.def_kind(def_id) {
+        DefKind::Ctor(..) | DefKind::Closure => return true,
+        DefKind::Fn | DefKind::AssocFn => {}
+        _ => return false,
+    }
+
+    // From this point on, it is valid to return true or false.
+    if tcx.sess.opts.unstable_opts.cross_crate_inline_threshold == InliningThreshold::Always {
+        return true;
+    }
+
     // Obey source annotations first; this is important because it means we can use
     // #[inline(never)] to force code generation.
     match codegen_fn_attrs.inline {
         InlineAttr::Never => return false,
         InlineAttr::Hint | InlineAttr::Always => return true,
         _ => {}
-    }
-
-    // This just reproduces the logic from Instance::requires_inline.
-    match tcx.def_kind(def_id) {
-        DefKind::Ctor(..) | DefKind::Closure => return true,
-        DefKind::Fn | DefKind::AssocFn => {}
-        _ => return false,
     }
 
     // Don't do any inference when incremental compilation is enabled; the additional inlining that

--- a/compiler/rustc_monomorphize/src/collector.rs
+++ b/compiler/rustc_monomorphize/src/collector.rs
@@ -176,6 +176,7 @@ use rustc_middle::mir::visit::Visitor as MirVisitor;
 use rustc_middle::mir::{self, Location};
 use rustc_middle::query::TyCtxtAt;
 use rustc_middle::ty::adjustment::{CustomCoerceUnsized, PointerCoercion};
+use rustc_middle::ty::layout::ValidityRequirement;
 use rustc_middle::ty::print::with_no_trimmed_paths;
 use rustc_middle::ty::{
     self, AssocKind, GenericParamDefKind, Instance, InstanceDef, Ty, TyCtxt, TypeFoldable,
@@ -921,6 +922,21 @@ fn visit_instance_use<'tcx>(
     debug!("visit_item_use({:?}, is_direct_call={:?})", instance, is_direct_call);
     if !should_codegen_locally(tcx, &instance) {
         return;
+    }
+
+    // The intrinsics assert_inhabited, assert_zero_valid, and assert_mem_uninitialized_valid will
+    // be lowered in codegen to nothing or a call to panic_nounwind. So if we encounter any
+    // of those intrinsics, we need to include a mono item for panic_nounwind, else we may try to
+    // codegen a call to that function without generating code for the function itself.
+    if let ty::InstanceDef::Intrinsic(def_id) = instance.def {
+        let name = tcx.item_name(def_id);
+        if let Some(_requirement) = ValidityRequirement::from_intrinsic(name) {
+            let def_id = tcx.lang_items().get(LangItem::PanicNounwind).unwrap();
+            let panic_instance = Instance::mono(tcx, def_id);
+            if should_codegen_locally(tcx, &panic_instance) {
+                output.push(create_fn_mono_item(tcx, panic_instance, source));
+            }
+        }
     }
 
     match instance.def {

--- a/compiler/rustc_monomorphize/src/partitioning.rs
+++ b/compiler/rustc_monomorphize/src/partitioning.rs
@@ -212,11 +212,17 @@ where
     let cgu_name_cache = &mut FxHashMap::default();
 
     for mono_item in mono_items {
-        // Handle only root items directly here. Inlined items are handled at
-        // the bottom of the loop based on reachability.
+        // Handle only root (GloballyShared) items directly here. Inlined (LocalCopy) items
+        // are handled at the bottom of the loop based on reachability, with one exception.
+        // The #[lang = "start"] item is the program entrypoint, so there are no calls to it in MIR.
+        // So even if its mode is LocalCopy, we need to treat it like a root.
         match mono_item.instantiation_mode(cx.tcx) {
             InstantiationMode::GloballyShared { .. } => {}
-            InstantiationMode::LocalCopy => continue,
+            InstantiationMode::LocalCopy => {
+                if Some(mono_item.def_id()) != cx.tcx.lang_items().start_fn() {
+                    continue;
+                }
+            }
         }
 
         let characteristic_def_id = characteristic_def_id_of_mono_item(cx.tcx, mono_item);

--- a/library/std/src/rt.rs
+++ b/library/std/src/rt.rs
@@ -155,7 +155,6 @@ fn lang_start_internal(
 }
 
 #[cfg(not(test))]
-#[inline(never)]
 #[lang = "start"]
 fn lang_start<T: crate::process::Termination + 'static>(
     main: fn() -> T,

